src/nvim/option.c

```diff
@@ -3738,6 +3738,10 @@ did_set_string_option (
       if (varp == &p_enc && *curbuf->b_p_keymap != NUL)
         (void)keymap_init();
 
+      if (varp == &p_enc) {
+        ui_update_encoding();
```

Comment:
Even if changing `&encoding` is weird, as long as it is allowed, it should not cause nvim to an inconsistent broken state where an arbitrary subset of utf-8 input is supported (åäö but not ÅÄÖ for instance),

Given the direction we're heading, the other option would be to _force_ `encoding=utf-8` (or always use utf-8 internally no matter its value), and leave both the detection and handling of non-utf8 terminals completely to the `tui` module.